RecoBTag/ONNXRuntime/plugins/ParticleTransformerAK4SonicJetTagsProducer.cc

```diff
@@ -0,0 +1,187 @@
+  const unsigned int target_n_vtx = std::clamp(max_n_vtx_counter, (unsigned int)1, (unsigned int)parT::n_sv_accept);
+
+  const std::map<parT::InputFeatures, unsigned int> target_n {
+    {parT::kChargedCandidates,     target_n_cpf},
+    {parT::kNeutralCandidates,     target_n_npf},
+    {parT::kVertices,              target_n_vtx},
+    {parT::kChargedCandidates4Vec, target_n_cpf},
+    {parT::kNeutralCandidates4Vec, target_n_npf},
+    {parT::kVertices4Vec,          target_n_vtx}
+  };
+ 
+  // loop through all groups of features
+  for (parT::InputFeatures ifeature = parT::kBegin; ifeature != parT::kEnd; ifeature = static_cast<parT::InputFeatures>(ifeature+1)) {
+    const auto &group_name = input_names_[ifeature];
+    auto &input = iInput.at(group_name);
+
+    input.setShape(0, target_n.at(ifeature));
+    auto tdata = input.allocate<float>(true);
+
+    for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+      const auto &taginfo = (*tag_infos)[jet_n];
+      const auto &features = taginfo.features();
+      auto &vdata = (*tdata)[jet_n];
+      if (ifeature == parT::kChargedCandidates || ifeature == parT::kChargedCandidates4Vec) 
+        parT_tensor_filler(vdata, ifeature, features.c_pf_features, target_n_cpf);
+      else if (ifeature == parT::kNeutralCandidates || ifeature == parT::kNeutralCandidates4Vec) 
+        parT_tensor_filler(vdata, ifeature, features.n_pf_features, target_n_npf);
+      else if (ifeature == parT::kVertices || ifeature == parT::kVertices4Vec)
+        parT_tensor_filler(vdata, ifeature, features.sv_features, target_n_vtx);
+    }
+    input.toServer(tdata);
+  }
+}
+
+void ParticleTransformerAK4SonicJetTagsProducer::produce(edm::Event& iEvent, const edm::EventSetup& iSetup, Output const &iOutput) {
+  edm::Handle<TagInfoCollection> tag_infos;
+  iEvent.getByToken(src_, tag_infos);
+
+  // initialize output collection
+  std::vector<std::unique_ptr<JetTagCollection>> output_tags;
+  if (!tag_infos->empty()) {
+    auto jet_ref = tag_infos->begin()->jet();
+    auto ref2prod = edm::makeRefToBaseProdFrom(jet_ref, iEvent);
+    for (std::size_t i = 0; i < flav_names_.size(); i++) {
+      output_tags.emplace_back(std::make_unique<JetTagCollection>(ref2prod));
+    }
+  } else {
+    for (std::size_t i = 0; i < flav_names_.size(); i++) {
+      output_tags.emplace_back(std::make_unique<JetTagCollection>());
+    }
+  }
+  if (!tag_infos->empty()) {
+    if (!skippedInference_) {
+      const auto &output1 = iOutput.begin()->second;
+      const auto &outputs_from_server = output1.fromServer<float>();
+
+      for (unsigned jet_n = 0; jet_n < tag_infos->size(); ++jet_n) {
+        const auto &taginfo = (*tag_infos)[jet_n];
+        const auto &jet_ref = tag_infos->at(jet_n).jet();
+
+        if (taginfo.features().is_filled) {
+          for (std::size_t flav_n = 0; flav_n < flav_names_.size(); flav_n++) {
+            (*(output_tags[flav_n]))[jet_ref] = outputs_from_server[jet_n][flav_n];
```

Comment:
this can be:
```cpp
(*(output_tags[flav_n]))[jet_ref] = taginfo.features().is_filled ? outputs_from_server[jet_n][flav_n] : -1.0;
```

so that the for loop doesn't have to be repeated